### ISSUE TITLE
MetadataName incorrect for imported VB namespace symbols

### ISSUE THREAD

**Issue by gafter** [dotnet/roslyn#26546]
`ISymbol.MetadataName` is specified to give the name for a symbol as it appears in metadata.  A class `ns.C` imported from metadata into VB cannot be distinguished from a class `NS.C` imported from the same metadata when viewed through the `INamespaceSymbol` API because they share the same enclosing namespace metadata name. One of these namespace names violates the contract for `MetadataName`, as it has a case that is different from its appearance in metadata.

I believe we should used a merged namespace symbol for this case so that each distinct type and its containers has a correctly-cased `MetadataName`.

This bug in the VB symbol API impacts the currently-proposed `ISymbolExtensions.IsAccessibleWithin` API.

**Comment by AlekseyTs** [comment]
This is not a bug, this is correct behavior from the language point of view. Namespaces that differ only by case represent the same **single** namespace that contains all types from all those namespaces. There is no concept of namespaces in metadata, therefore, MetadataName simply returns the Name. This is true for any other symbol that doesn't have matching concept in metadata. 
I agree that today the information about the actual casing is hard to get. I think it can be obtained from SymbolDisplay for a type, but that is certainly on the most convenient way. We probably need to add a more convenient API for that.

**Comment by AlekseyTs** [comment]
In metadata, namespace name is a property of a type declaration. So, I suggest to follow the same pattern and adding a MetadataNamespaceName to the INamedTypeSymbol, this is how VB tracks this name internally, BTW. This will work equally well for both languages, an implementation will differ, of course.

**Comment by gafter** [comment]
We added Roslyn's concept of `INamespaceSymbol` to extend the metadata model into our symbol model. An `INamedTypeSymbol` already has a containing namespace in the symbol model. `INamespaceSymbol` already has a `MetadataName` property. It feels wrong to add more information about the namespace into the named type symbol when the current API is fully capable of expressing the situation.

**Comment by AlekseyTs** [comment]
I can see how you can feel this way. However, INamespaceSymbol **reflects the language view** and simply doesn't have a direct mapping to metadata. The MetadataName property was added much later and it is relevant only to some symbols that actually have declarations in metadata, I already explained earlier the rational behind this and why it is on ISymbol. The behavior of that property on INamespaceSymbol is by design, it behaves how it was intended to behave, perhaps it is not clearly documented and that certainly can lead to some confusion. I understand that the information is missing and we should remedy this, but changing behavior of MetadataName on INamespaceSymbol is not the right way to do this. It is going to violate our primary principle - symbols reflect language view of the world, not metadata view of the world.

**Comment by gafter** [comment]
> I understand that the information is missing and we should remedy this, but changing behavior of MetadataName on INamespaceSymbol is not the right way to do this.

I disagree. I understand the design errors that led to our current situation, and I believe they are best addressed directly rather than by slavishly following the previous design rationale. I do not believe that the hypothetical departure from the language view of the world is one that would impact clients that use the APIs as designed.

**Comment by AlekseyTs** [comment]
> I do not believe that the hypothetical departure from the language view of the world is one that would impact clients that use the APIs as designed.

There is one detail though. The MetadataName API was never designed to work on INamespaceSymbol the way you assume it was designed. The implementation of ISymbolExtensions.IsAccessibleWithin API, in particular, is a good example of using API as not designed and that is why it fails. A desire for an API to work in a certain way should not be confused with designed behavior. One could advocate for a change in design, but I do not believe this is warranted in this case because documentation can easily avoid confusion in the future and violation of **our primary principle - symbols reflect language view of the world, not metadata view of the world** is going to impact clients.

**Comment by gafter** [comment]
The proposed use of merged namespaces for this situation does not violate the language view any more than it violates that language view where merged namespaces are used elsewhere today. To be impacted by that proposed change, a client would need to be using the APIs improperly (i.e. not starting at the root namespace to look up names).

**Comment by AlekseyTs** [comment]
> The proposed use of merged namespaces for this situation does not violate the language view any more than it violates that language view where merged namespaces are used elsewhere today. 

But it does. This was discussed at length in details at the Roslyn design meeting (you were there by the way) and the consensus was that from VB language point of view, regardless of how many different spelling are used to declare namespaces in source and how many different namespace spellings are used by types in metadata, there is only one namespace declared per unit of compilation (i.e. module) and that should not be represented by any merged namespace symbol, but should be represented by the only non-merged namespace symbol. It was not easy to implement that design decision, but the decision was that that was worth the cost.

> To be impacted by that proposed change, a client would need to be using the APIs improperly (i.e. not starting at the root namespace to look up names).

I see nothing improper for a VB user to go up from type to its containing namespace and expect to find all members of that namespace according to the language rules that are declared in the same unit of compilation without a need to go up to the compilation and then back down.
